Add ModelManager to schedule calls between different model backend

RomanPryima1 opened the pull request:
Add ModelManager to schedule calls between different model backend

## Description
Introduce a ModelManger instance aimed to spread a load among provided set of models.
Currently this manager uses only one spreading strategy "round robin".
It's intended to be extended with more strategies, once they are defined.

## Motivation and Context

Why is this change required? What problem does it solve?
 `close #1031` 

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

Wendong-Fan commented on camel/models/base_model.py:
```diff
@@ -138,3 +142,83 @@ def stream(self) -> bool:
+            self.scheduling_strategy = self.round_robin
+
+    @property
+    def model_type(self):
+        """Return type of the first model.
+        Assuming that all models instances are the same type."""
+        return self.models[0].model_type if self.models else None
```
actually we shouldn't make this assumption, `model_type` could be vary , same as other properties

RomanPryima1 replied:
Changed according to this requirement.

Wendong-Fan commented on camel/models/base_model.py:
```diff
@@ -138,3 +142,83 @@ def stream(self) -> bool:
         return False
+
+
+class ModelManager:
+    def __init__(
```
I think we can make this class under a new python file called `model_manager.py`

RomanPryima1 replied:
Moved to a separate module.

Wendong-Fan reviewed (commented):
Thanks for the contribution, Left some comments below, I think some key componets need to be added
1. we should use index to track the model currently is running, and let user know the model has been switched
2. the reason why we switch the model is due to the original model couldn't been used, but in the code there's no check for this
3. we need add more `scheduling_strategy `, from the issue description we have mentioned 2, `round-robin` and `always_first_model `, it would be better if we also support more like `random` etc..
4. lacking handle for the message as described in the issue
5. need to write example code to showcase the usage under `examples/models`

willshang76 commented on camel/models/base_model.py:
```diff
@@ -138,3 +142,83 @@ def stream(self) -> bool:
+        models: Union[BaseModelBackend, List[BaseModelBackend]],
+        scheduling_strategy: str = "round_robin",
+    ):
+        """
```
Please make sure the format is consistant with existing code. 
1) using r""" for comment
2) Using indents for comment content like https://github.com/camel-ai/camel/blob/master/camel/models/base_model.py#L30-L47
3) The comment should be above the __init__ method def

RomanPryima1 replied:
Rewrote the docstring.

RomanPryima1 commented:
> Thanks for the contribution, Left some comments below, I think some key componets need to be added
> 
> 1. we should use index to track the model currently is running, and let user know the model has been switched
> 2. the reason why we switch the model is due to the original model couldn't been used, but in the code there's no check for this
> 3. we need add more `scheduling_strategy `, from the issue description we have mentioned 2, `round-robin` and `always_first_model `, it would be better if we also support more like `random` etc..
> 4. lacking handle for the message as described in the issue
> 5. need to write example code to showcase the usage under `examples/models`

1. Added logging for both cases: model finished successfully and model failed.
2. Added try/except closure for models switching.
3. Added "random_model" and "always_first" strategies. The last one falls into "round_robin" in case the first model failed during processing.
4. Added logging
5. Added example to "examples/models/openai_compatibility_model_examples/model_manger.py" in the way I understood how it should be written.

willshang76 commented:
Please sync the latest master branch. Thanks!

willshang76 commented on camel/agents/chat_agent.py:
```diff
@@ -79,6 +79,10 @@
 
 
+class ModelProcessingError(Exception):
+    """Raised when an error occurs during model processing."""
```
Please update the docstring by following the unified format we are using today. Thanks

RomanPryima1 replied:
Updated

willshang76 commented on examples/models/openai_compatibility_model_examples/model_manger.py:
```diff
@@ -0,0 +1,55 @@
+
+assistant_sys_msg = "Testing two models in ModelManager"
+
+agent = ChatAgent(assistant_sys_msg, model=[model1, model2])
```
I believe that we should expose the capability for the users to change the schedule strategy, please update and include an example here to show how to change the strategy. 
We could define an enum type here(https://github.com/camel-ai/camel/blob/master/camel/types/enums.py) and map the enum type to the strategy in the model manager. In the chatagent, expose a param for the user to set the schedule strategy.

RomanPryima1 replied:
Extended an example. But I would avoid using Enum for scheduling_strategy param, as we allow user to use his own strategy implementation with own name.

willshang76 commented on camel/agents/chat_agent.py:
```diff
@@ -943,8 +949,31 @@ def _step_model_response(
+        response = None
         # Obtain the model's response
-        response = self.model_backend.run(openai_messages)
+        for _ in range(len(self.model_backend.models)):
```
Please help me understand, why would we run all the models for every message?

RomanPryima1 replied:
Added break statement for fixing this bug.

Wendong-Fan reviewed (commented):
Thanks @RomanPryima1 ! Overall looks great, left some comments below

RomanPryima1 commented on camel/models/model_manager.py:
```diff
@@ -0,0 +1,122 @@
+        try:
+            response = self.current_model.run(messages)
+        except Exception as exc:
+            if self.scheduling_strategy == self.always_first:
+                self.scheduling_strategy = self.round_robin
```
Added logging

RomanPryima1 commented on camel/models/model_manager.py:
```diff
@@ -0,0 +1,122 @@
+logger = logging.getLogger(__name__)
+
+
+class ModelManager:
```
Added such functionality.

Wendong-Fan replied:
seems here only parameter added

RomanPryima1 commented on camel/models/model_manager.py:
```diff
@@ -0,0 +1,122 @@
+    @property
+    def model_config_dict(self):
+        """Return model_config_dict of the current model."""
+        return self.current_model.model_config_dict
```
Updated

RomanPryima1 commented on camel/agents/chat_agent.py:
```diff
@@ -943,8 +949,31 @@ def _step_model_response(
         # Obtain the model's response
-        response = self.model_backend.run(openai_messages)
+        for _ in range(len(self.model_backend.models)):
+            try:
+                response = self.model_backend.run(openai_messages)
```
fixed

RomanPryima1 commented on camel/models/model_manager.py:
```diff
@@ -0,0 +1,122 @@
+    def __init__(
+        self,
+        models: Union[BaseModelBackend, List[BaseModelBackend]],
+        scheduling_strategy: str = "round_robin",
```
Literal limits possible options. But once we add a possibility of extension with custom strategies, it can have any string value.

Wendong-Fan commented on camel/agents/chat_agent.py:
```diff
@@ -79,6 +80,10 @@
 
+class ModelProcessingError(Exception):
+    r"""Raised when an error occurs during model processing."""
+
```
move this under model_manager.py would be better

Wendong-Fan commented on camel/models/model_manager.py:
```diff
@@ -0,0 +1,220 @@
+    ):
+        scheduling_strategy = scheduling_strategy or "round_robin"
+        if TYPE_CHECKING:
+            assert isinstance(scheduling_strategy, str)
```
seems code here could be simpified

Wendong-Fan commented on camel/models/model_manager.py:
```diff
@@ -0,0 +1,220 @@
+        custom_strategy Optional[Callable]: Scheduling strategy method
+            provided by user in case when none of existent strategies fits.
+            When custom strategy is provided, it will be used
+            disrespectfully to "scheduling_strategy" parameter.
```
docstring need to be removed

Wendong-Fan commented on camel/models/model_manager.py:
```diff
@@ -0,0 +1,220 @@
+        """
+        if not callable(strategy_fn):
+            raise ValueError("strategy_fn must be a callable function.")
+        setattr(self, name, strategy_fn)
```
seems one bug here, otherwise the example code would not able to run
```suggestion
        setattr(self, name, strategy_fn.__get__(self))
```

Wendong-Fan commented on camel/models/model_manager.py:
```diff
@@ -0,0 +1,220 @@
+        self.current_model.model_config_dict = model_config_dict
+
+    @property
+    def current_model_index(self) -> Optional[int]:
```
seems it will never return None

Wendong-Fan reviewed (commented):
Thanks @RomanPryima1 , left some comments below and added one more commit based on the review here: https://github.com/camel-ai/camel/pull/1190/commits/e2da38f4d2c4c3a2a1b8ccbe7d1f99594656de7c, feel free to check and leave your comment~
